AppConfiguration 2024-09-01-preview API version

hahahahahaiyiwen opened the pull request:
# ARM (Control Plane) API Specification Update Pull Request 

> [!TIP]
> Overwhelmed by all this guidance? See the `Getting help` section at the bottom of this PR description.

## PR review workflow diagram

Please understand this diagram before proceeding. It explains how to get your PR approved & merged.

![spec_pr_review_workflow_diagram](https://github.com/Azure/azure-rest-api-specs/assets/4429827/5bb5e7ce-8aff-4dbb-a3f8-0d9b68fef5b1)

## Purpose of this PR

What's the purpose of this PR? Check the specific option that applies. This is **mandatory**!

  - [ ] New resource provider.
  - [x] New API version for an existing resource provider. (If API spec is not defined in TypeSpec, the PR should have been created in adherence to [OpenAPI specs PR creation guidance](https://aka.ms/azsdkdocs/createopenapispec)).
  - [ ] Update existing version for a new feature. (This is applicable only when you are revising a private preview API version.)
  - [ ] Update existing version to fix OpenAPI spec quality issues in S360.
  - [ ] Other, please clarify:
    - _edit this with your clarification_

## Due diligence checklist

To merge this PR, you **must** go through the following checklist and confirm you understood 
and followed the instructions by checking all the boxes:

- [x] I confirm this PR is modifying Azure Resource Manager (ARM) related specifications, and not data plane related specifications.
- [x] I have reviewed following [Resource Provider guidelines](https://aka.ms/rpguidelines), including
  [ARM resource provider contract](https://aka.ms/azurerpc) and
  [REST guidelines](https://github.com/microsoft/api-guidelines/blob/vNext/azure/Guidelines.md) (estimated time: 4 hours).  
  I understand this is required before I can proceed to the diagram Step 2, "ARM API changes review", for this PR.

## Additional information

<details>
<summary> Viewing API changes</summary>

For convenient view of the API changes made by this PR, refer to the URLs provided in the table 
in the `Generated ApiView` comment added to this PR. You can use ApiView to show API versions diff. 

</details>
<details>
<summary>Suppressing failures</summary>

If one or multiple validation error/warning suppression(s) is detected in your PR, please follow the 
[suppressions guide](https://aka.ms/azsdk/pr-suppressions) to get approval.

</details>

## Getting help

- First, please carefully read through this PR description, from top to bottom. Please fill out the `Purpose of this PR` and `Due diligence checklist`.
- If you don't have permissions to remove or add labels to the PR, request `write access` per [aka.ms/azsdk/access#request-access-to-rest-api-or-sdk-repositories](https://aka.ms/azsdk/access#request-access-to-rest-api-or-sdk-repositories")
- To understand what you must do next to merge this PR, see the `Next Steps to Merge` comment. It will appear within few minutes of submitting this PR and will continue to be up-to-date with current PR state.
- For guidance on fixing this PR CI check failures, see the hyperlinks provided in given failure 
  and https://aka.ms/ci-fix.
- For help with ARM review (PR workflow diagram Step 2), see https://aka.ms/azsdk/pr-arm-review.
- If the PR CI checks appear to be stuck in `queued` state, please add a comment with contents `/azp run`.
  This should result in a new comment denoting a `PR validation pipeline` has started and the checks should be updated after few minutes.
- If the help provided by the previous points is not enough, post to https://aka.ms/azsdk/support/specreview-channel and link to this PR.

hahahahahaiyiwen committed:
Add 2024-09-01-preview

hahahahahaiyiwen committed:
Fix validation error, add permission

hahahahahaiyiwen committed:
Fix validation

hahahahahaiyiwen committed:
Remove permission and update example

hahahahahaiyiwen committed:
Merge branch 'Azure:main' into appconfig/add2024-09-01-preview

avanigupta commented on specification/appconfiguration/resource-manager/Microsoft.AppConfiguration/preview/2024-09-01-preview/appconfiguration.json:
```diff
@@ -0,0 +1,3168 @@
+          "description": "The kind of the SAS token.",
+          "type": "string",
+          "enum": [
+            "Primary",
```
Are these case sensitive? They shouldnt need to be.

hahahahahaiyiwen replied:
I think generally we make the value of enum that modeled as string in this case sensitive format. For example, "Enabled", "Disabled" for "publicNetworkAccess"

hahahahahaiyiwen replied:
For resourceType "kv" and "snapshot", if following the same convention, maybe I should also update them to "Kv" and "Snapshot"

avanigupta replied:
So what happens if someone provides 'enabled' for "publicNetworkAccess"? Do we fail the request?

hahahahahaiyiwen replied:
No, we didn't fail the request, the allowed values check of "publicNetworkAccess" ignores cases. But we did it differently for different properties. For "dataplaneproxy.authenticationMode", the check doesn't ignore case and we will fail the request.

avanigupta replied:
Ok, so the case sensitivity is decided in the implementation. We can ignore the case for SAS parameters in the RP implementation.

avanigupta commented on specification/appconfiguration/resource-manager/Microsoft.AppConfiguration/preview/2024-09-01-preview/appconfiguration.json:
```diff
@@ -0,0 +1,3168 @@
+          "format": "date-time"
+        },
+        "cch":{
+          "description": "Time (in seconds) for which the response may be cached by clients.",
```
```suggestion
          "description": "Time (in seconds) for which the data plane response may be cached by clients. App Config service sets the Cache-Control response header `max-age` to the value that's specified on the SAS token. See [rfc9111](https://www.rfc-editor.org/rfc/rfc9111#name-max-age-2) for more details.",
```

avanigupta commented on specification/appconfiguration/resource-manager/Microsoft.AppConfiguration/preview/2024-09-01-preview/appconfiguration.json:
```diff
@@ -0,0 +1,3168 @@
+          "description": "Am array of tag filters used to match tags.",
+          "type": "array",
+          "items": {
+            "description": "A filter used to match tag.",
```
```suggestion
            "description": "A filter used to match key-values by their tags, specified in `tagName=tagValue` format.",
```

avanigupta commented on specification/appconfiguration/resource-manager/Microsoft.AppConfiguration/preview/2024-09-01-preview/appconfiguration.json:
```diff
@@ -0,0 +1,3168 @@
+          }
+        },
+        "name":{
+          "description": "The name of the snapshot. This is only applicable when the resourceType is snapshot.",
```
Can you add the same description about resourceType restriction in key/label/tags

avanigupta commented on specification/appconfiguration/resource-manager/Microsoft.AppConfiguration/preview/2024-09-01-preview/appconfiguration.json:
```diff
@@ -0,0 +1,3168 @@
+      "description": "The result of a request to generate a SAS token.",
+      "type": "object",
+      "properties": {
+        "tokenValue":{
```
Does the order of properties defined here matter? If yes, we should put tokenValue at the end and other fields in the same order as GenerateSasTokenParameters

hahahahahaiyiwen replied:
I don't think it matters but I moved "tokenValue" to the end as you suggested.

avanigupta commented on specification/appconfiguration/resource-manager/Microsoft.AppConfiguration/preview/2024-09-01-preview/examples/ConfigurationStoresGenerateSasToken.json:
```diff
@@ -0,0 +1,41 @@
+  "responses": {
+    "200": {
+      "body": {
+        "tokenValue": "sas=res=kv;sp=key,label,tags;exp=2024-08-01T00:00:00Z;cch=3600;oid=<ObjectId>;tid=<TenantId>;p=read;sig=<hmac256signature>",
```
```suggestion
        "tokenValue": "sas=res=kv;sp=key,label,tags;exp=2024-09-01T00:00:00Z;cch=3600;oid=<ObjectId>;tid=<TenantId>;p=read;sig=<hmac256signature>",
```

avanigupta commented on specification/appconfiguration/resource-manager/Microsoft.AppConfiguration/preview/2024-09-01-preview/appconfiguration.json:
```diff
@@ -0,0 +1,3168 @@
+        ],
+        "responses": {
+          "200": {
+            "description": "The request was successful; the request was well-formed and received properly."
```
Is it odd that this API returns no response body?

hahahahahaiyiwen replied:
I think it's fine for a "POST" operation. For example, purge post operation also returns no response body.

hahahahahaiyiwen committed:
Update specification/appconfiguration/resource-manager/Microsoft.AppConfiguration/preview/2024-09-01-preview/appconfiguration.json

Co-authored-by: Avani Gupta <avanigupta@users.noreply.github.com>

hahahahahaiyiwen committed:
Update specification/appconfiguration/resource-manager/Microsoft.AppConfiguration/preview/2024-09-01-preview/appconfiguration.json

Co-authored-by: Avani Gupta <avanigupta@users.noreply.github.com>

hahahahahaiyiwen committed:
Address comments

avanigupta commented on specification/appconfiguration/resource-manager/Microsoft.AppConfiguration/preview/2024-09-01-preview/appconfiguration.json:
```diff
@@ -0,0 +1,3168 @@
+          "readOnly": true
+        },
+        "cch":{
+          "description": "Time (in seconds) for which the response may be cached by clients.",
```
Same description update for cch here

hahahahahaiyiwen committed:
Address comments

jimmyca15 commented on specification/appconfiguration/resource-manager/Microsoft.AppConfiguration/preview/2024-09-01-preview/appconfiguration.json:
```diff
@@ -0,0 +1,3168 @@
+        "tags": [
+          "ConfigurationStores"
+        ],
+        "description": "Generates the SAS token for scoped, read-only access of the specified configuration store.",
```
```suggestion
        "description": "Generates a SAS token for scoped, read-only access of the specified configuration store.",
```

jimmyca15 commented on specification/appconfiguration/resource-manager/Microsoft.AppConfiguration/preview/2024-09-01-preview/appconfiguration.json:
```diff
@@ -0,0 +1,3168 @@
+            "description": "The object containing information for the SAS token generation request.",
+            "required": true,
+            "schema": {
+              "$ref": "#/definitions/GenerateSasTokenParameters"
```
```suggestion
              "$ref": "#/definitions/SasTokenGenerationParameters"
```

jimmyca15 commented on specification/appconfiguration/resource-manager/Microsoft.AppConfiguration/preview/2024-09-01-preview/appconfiguration.json:
```diff
@@ -0,0 +1,3168 @@
+          "200": {
+            "description": "The request was successful; the request was well-formed and received properly.",
+            "schema": {
+              "$ref": "#/definitions/SasTokenGenerateResult"
```
```suggestion
              "$ref": "#/definitions/SasTokenGenerationResult"
```

jimmyca15 commented on specification/appconfiguration/resource-manager/Microsoft.AppConfiguration/preview/2024-09-01-preview/appconfiguration.json:
```diff
@@ -0,0 +1,3168 @@
+        }
+      }
+    },
+    "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.AppConfiguration/configurationStores/{configStoreName}/resetSas": {
```
Isn't this supposed to be reset signing key? Not reset sas?

jimmyca15 replied:
The term SAS is a bit incomplete. SAS token is more complete. I'd suggest the following API update.

```suggestion
    "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.AppConfiguration/configurationStores/{configStoreName}/resetSasTokens": {
```

cc @avanigupta

jimmyca15 commented on specification/appconfiguration/resource-manager/Microsoft.AppConfiguration/preview/2024-09-01-preview/appconfiguration.json:
```diff
@@ -0,0 +1,3168 @@
+          "type": "string",
+          "enum": [
+            "Primary",
+            "Secondary"
```
I'm not aware of there being primary and secondary SAS tokens. Only primary and secondary signing keys.

jimmyca15 replied:
@avanigupta

jimmyca15 replied:
I found the design doc updates. Deleting this.

jimmyca15 commented on specification/appconfiguration/resource-manager/Microsoft.AppConfiguration/preview/2024-09-01-preview/appconfiguration.json:
```diff
@@ -0,0 +1,3168 @@
+          "type": "string",
+          "format": "date-time"
+        },
+        "cch":{
```
What is cch? cache control header? Is there a reason to abbreviate?

@avanigupta

avanigupta replied:
Yes, this is a field in the SAS token value. Abbreviated to keep the token as short as possible. Storage uses [rscc ](https://learn.microsoft.com/en-us/rest/api/storageservices/create-user-delegation-sas#specify-query-parameters-to-override-response-headers) for this field.

jimmyca15 replied:
It seems to me that for the RP API this should either be expanded to `cacheControlHeader` or `cacheControlMaxAge`. Out of the two, I think cacheControlMaxAge would be more helpful. And if there's an agreeance on that I would suggest changing the sas token parameter key to `ccma`

avanigupta replied:
cacheControlMaxAge and ccma sound good

jimmyca15 commented on specification/appconfiguration/resource-manager/Microsoft.AppConfiguration/preview/2024-09-01-preview/appconfiguration.json:
```diff
@@ -0,0 +1,3168 @@
+          "format": "date-time"
+        },
+        "cch":{
+          "description": "Time (in seconds) for which the data plane response may be cached by clients. App Config service sets the Cache-Control response header `max-age` to the value that's specified on the SAS token. See [rfc9111](https://www.rfc-editor.org/rfc/rfc9111#name-max-age-2) for more details.",
```
```suggestion
          "description": "Time (in seconds) for which the data plane response may be cached by clients. App Configuration sets the Cache-Control response header `max-age` to the value that's specified on the SAS token. See [rfc9111](https://www.rfc-editor.org/rfc/rfc9111#name-max-age-2) for more details.",
```

jimmyca15 commented on specification/appconfiguration/resource-manager/Microsoft.AppConfiguration/preview/2024-09-01-preview/appconfiguration.json:
```diff
@@ -0,0 +1,3168 @@
+        "tags": [
+          "ConfigurationStores"
+        ],
+        "description": "Reset the SAS token of a given type.",
```
```suggestion
        "description": "Reset all SAS tokens of a given kind (primary/secondary).",
```

jimmyca15 commented on specification/appconfiguration/resource-manager/Microsoft.AppConfiguration/preview/2024-09-01-preview/appconfiguration.json:
```diff
@@ -0,0 +1,3168 @@
+          "description": "The parameters to define the data plane resource scope.",
+          "$ref": "#/definitions/SignedParameters"
+        },
+        "exp":{
```
I would expect full names for ARM API.

avanigupta replied:
Are you suggesting we accept full names in request body and abbreviate as needed while generating the token?

jimmyca15 replied:
Yes.

avanigupta replied:
Okay, so we can change these three request paramters:
- exp => expires
- cch => cache_control_max_age
- sas_kind => kind

hahahahahaiyiwen replied:
Got it

jimmyca15 replied:
How about expiration instead of expires?

avanigupta replied:
We use `expires` for snapshots: 
https://github.com/Azure/azure-rest-api-specs/blob/0117e474a03d0b5b4cdd3b713105ae17196c18ce/specification/appconfiguration/data-plane/Microsoft.AppConfiguration/stable/2023-11-01/appconfiguration.json#L2978

jimmyca15 replied:
Ah, okay thank you. I thought we must have a precedent. expires it is then.

hahahahahaiyiwen replied:
For resourceType value, should we keep "kv" or update to "keyValue"?

jimmyca15 replied:
Keep `kv`. The concern I was raising was purely over property names.

jimmyca15 commented on specification/appconfiguration/resource-manager/Microsoft.AppConfiguration/preview/2024-09-01-preview/appconfiguration.json:
```diff
@@ -0,0 +1,3168 @@
+          "description": "Time (in seconds) for which the data plane response may be cached by clients. App Config service sets the Cache-Control response header `max-age` to the value that's specified on the SAS token. See [rfc9111](https://www.rfc-editor.org/rfc/rfc9111#name-max-age-2) for more details.",
+          "type": "number"
+        },
+        "sasKind":{
```
Is `kind` not enough?

avanigupta replied:
`kind` sounds good. We can update it in the request body for both `generateSasToken` and `resetSasTokens` APIs

jimmyca15 commented on specification/appconfiguration/resource-manager/Microsoft.AppConfiguration/preview/2024-09-01-preview/appconfiguration.json:
```diff
@@ -0,0 +1,3168 @@
+          "type": "string",
+          "readOnly": true
+        },
+        "tokenValue":{
```
Can we just use `value`?

avanigupta replied:
Sure

hahahahahaiyiwen committed:
Update specification/appconfiguration/resource-manager/Microsoft.AppConfiguration/preview/2024-09-01-preview/appconfiguration.json

Co-authored-by: Jimmy Campbell <jimmyca@microsoft.com>

hahahahahaiyiwen committed:
Address comments

hahahahahaiyiwen committed:
Address comments

hahahahahaiyiwen committed:
Address comments

hahahahahaiyiwen committed:
Address comments

hahahahahaiyiwen committed:
Address comments

hahahahahaiyiwen committed:
Merge branch 'Azure:main' into appconfig/add2024-09-01-preview

XiaofeiCao committed:
java update automation config (#29777)

* java update automation config

* format

msyyc committed:
rename property name for python sdk (#29793)

zhuoyunwang committed:
[Oracle Database@Azure] Add new preview and stable API versions for 2024-06-01 (#29585)

* add new preview and stable API versions

* remove unnecessary add decorator

* re-compile after merging with main

---------

Co-authored-by: zhuoyun wang <zhuoyunw@zhuoyuns-Virtual-Machine.local>

mikeharder committed:
[vmware] Add ".Management" suffix to TypeSpec folder name (#29748)

mikeharder committed:
[Informatica.DataManagement] Add .Management suffix to TypeSpec folder name (#29738)

JoshLove-msft committed:
Fix acronym casing (#29797)

* Fix acronym casing

* tsp format

mentat9 committed:
Add Typespec conversion option to intake template (#29616)

* Add Typespec converstion option to intake template

Intake form needs an option to indicate the author is converting swagger to Typespec. This change adds a checkbox for that option and indicates it should not be combined with adding a new API version.

* Update control_plane_template.md

Update text and add link to TypeSpec conversion docs per feedback.

* Update control_plane_template.md

* Update control_plane_template.md

Update help link for TypeSpec conversion

---------

Co-authored-by: Konrad Jamrozik <kojamroz@microsoft.com>

KesemSharabi committed:
Update powerbidedicated.json (#29322)

Updated based on this doc bug:
https://dev.azure.com/msft-skilling/Content/_workitems/edit/154565

mikeharder committed:
Add tsp-client to root package.json (#29803)

- Allows devs to run `npx tsp-client` after `npm ci`, with no additional installs
- Pins versions of `tsp-client` and its deps in the root package-lock.json
- Fixes #29686

bexxx committed:
Add 2024-05-15-preview of SpeechToText API (CognitiveServices) (#29804)

* Add preview GA as base

* Updates for new preview version

* Update readme.md

* Added required ttl to examples

* Add description

* Add missing example

* Keep latest GA latest

madhura-np commented:
A required check has failed - please fix. 
https://github.com/Azure/azure-rest-api-specs/pull/29677/checks?check_run_id=27297663464

madhura-np commented:
Please ensure the first iteration contains the base-api version as is, according to the guidelines here - https://armwiki.azurewebsites.net/rp_onboarding/process/api_review.html
This helps reviewers only review the updates to new versions and not have to review the full API files.

hahahahahaiyiwen committed:
Add 2024-09-01-preview by copying 2023-09-01-preview

hahahahahaiyiwen committed:
Merge
